index.rst

```diff
@@ -672,43 +701,61 @@ REST API
 To facilitate debugging of prepuller issues, there is a read-only REST API to see the status of a prepull configuration.
 Changing the configuration requires changing the Helm chart or the generated ``ConfigMap`` object and restarting the lab controller.
 
-All of these API calls require ``admin:notebook`` scope.
+All of these API calls require ``admin:jupyterlab`` scope.
 
 ``GET /nublado/spawner/v1/images``
-    Returns a list of known images and their names.
+    Returns an object consisting of the images allowed by the bot users,
+    followed by a key ``all`` whose contents are all known images and
+    their names.  Each of the top-level object keys is the string
+    literal exactly as shown.
     This parses the available images according to SQR-059_ and shows the results:
 
     .. code-block:: json
 
        {
            "recommended": {
-               "url": "<full image url>",
-               "tag": "<image tag>",
+               "path": "<full Docker image path>",
+               "tags": {
+                   "<image tag>": "<human readable name>"
+               },
                "name": "<human readable name>",
+	       "digest": "<image digest>",
```

Comment:
Not sure why we want to return this. What client would use this information?

Reply:
Oh, I think I see, this is for debugging cases where we've moved the tag and the image cache may not be up to date.

Reply:
So, yes, that's my primary use case, because what I've cared about in the prepuller.  It's also enabled us to keep pushing image digests into the spawned containers (this used to require K8s discovery from the spawned container, but we no longer grant any K8s privileges to our user Lab), because we can always just ask for the spawned pod by name, tag, and digest, and therefore it's available in the image description within the pod.

But that brings us an interesting further optimization we can make.  In the current Jupyterlab controller design, we will always have the digest.  I had a realization about when mobu or noteburst wants to pull an image by just tag and name: we know, definitionally, *all images*.  We can fail early by failing to match the tag from the "all images" list--which also contains all their digests, so we (assuming we find the tag in the list) know the matching digest (and can thus ask for it in the pull, and know what the digest is within our container), and if we do NOT resolve our desired tag to something in the "all images" list, then we can fail before ever trying to spawn anything: we don't have to wait until the spawned pod crashes with "ImagePullBackoff" because the image is not available on the repository.  (I haven't implemented that yet...but I can do so easily.)